test/lock/pessimistic_lock_test.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2021 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "lock/pessimistic_lock.hpp"
+
+#include <thread>
+#include <vector>
+
+#include "common.hpp"
+#include "gtest/gtest.h"
+
+namespace dbgroup::lock::test
+{
+class PessimisticLockFixture : public ::testing::Test
+{
+ protected:
+  /*####################################################################################
+   * Setup/Teardown
+   *##################################################################################*/
+
+  void
+  SetUp() override
+  {
+  }
+
+  void
+  TearDown() override
+  {
+  }
+
+  /*####################################################################################
+   * Functions for verification
+   *##################################################################################*/
+
+  void
+  VerifySharedLockWithMultiThread()
+  {
+    EXPECT_EQ(lock_.GetLock(), 0);
+
+    LockSharedWithMultiThread();
+    EXPECT_EQ(lock_.GetLock(), 2 * kThreadNum);
+
+    UnlockSharedWithMultiThread();
+    EXPECT_EQ(lock_.GetLock(), 0);
+  }
+
+  void
+  VerifyLockWithSingleThread()
+  {
+    EXPECT_EQ(lock_.GetLock(), 0);
+
+    lock_.Lock();
+    EXPECT_EQ(lock_.GetLock(), 1);
+
+    lock_.Unlock();
+    EXPECT_EQ(lock_.GetLock(), 0);
+  }
+
+  /*####################################################################################
+   * Public utility functions
+   *##################################################################################*/
+  void
```

Comment:
ロックについては，実際にブロックが行われるか確認しておいた方がよいと思います．以下はざっと書いてみたサンプルで，`lock_.Lock()`実行後に`VerifyLockShared(kExpectFailed)`みたいに呼び出す想定です（`constexpr bool kExpectFailed = false`）．別スレッドでロックを取りに行って，1ms経ってもロックが取れなければ失敗とみなします．

```suggestion
  void
  VerifyLockShared(const bool expect_success)
  {
    // try to get a shared lock by another thread
    auto lock_shared = [this] (std::promise<void> p) {
      lock_.LockShared();
      p.set_value();
    };
    std::promise<void> p{};
    auto&& f = p.get_future();
    std::thread t{lock_shared, std::move(p)};

    // after one millisecond has passed, give up on acquiring the lock
    const auto rc = f.wait_for(std::chrono::milliseconds{1});

    // verify status to check locking is succeeded
    if (expect_success) {
      ASSERT_EQ(rc, std::future_status::ready);
    } else {
      ASSERT_EQ(rc, std::future_status::timeout);
      lock_.UnLock();  // unlock to join the thread
    }
    t.join();
  }

  void
```

面倒なのは最後の`lock_.UnLock()`で，立ち上げたスレッドを回収するためにロックを外す必要があります．排他ロックの方を試すときは`std::atomic_size_t s_lock_count`で現在の共有ロック（排他ロックも？）の数を把握しておいて，whileか何かですべての共有ロックを解放するまでループを回す必要があると思います．

Reply:
ありがとうございます！！
(こんな風に実装できるんですね...)